server/src-lib/Hasura/Server/Auth/WebHook.hs

```diff
@@ -0,0 +1,150 @@
+  ) where
+
+import           Control.Exception         (try)
+import           Control.Lens
+import           Control.Monad.Trans.Maybe
+import           Data.Aeson
+import           Data.Time.Clock           (UTCTime, addUTCTime, getCurrentTime)
+import           Hasura.Server.Version     (HasVersion)
+
+import qualified Data.Aeson                as J
+import qualified Data.ByteString.Lazy      as BL
+import qualified Data.HashMap.Strict       as Map
+import qualified Data.Text                 as T
+import qualified Network.HTTP.Client       as H
+import qualified Network.HTTP.Types        as N
+import qualified Network.Wreq              as Wreq
+
+import           Data.Parser.CacheControl
+import           Data.Parser.Expires
+import           Hasura.HTTP
+import           Hasura.Logging
+import           Hasura.Prelude
+import           Hasura.RQL.Types
+import           Hasura.Server.Logging
+import           Hasura.Server.Utils
+
+
+data AuthHookType
+  = AHTGet
+  | AHTPost
+  deriving (Eq)
+
+instance Show AuthHookType where
+  show AHTGet  = "GET"
+  show AHTPost = "POST"
+
+isPost :: AuthHookType -> Bool
+isPost AHTGet  = False
+isPost AHTPost = True
+
+
+data AuthHookG a b
+  = AuthHookG
+  { ahUrl  :: !a
+  , ahType :: !b
+  } deriving (Show, Eq)
+
+type AuthHook = AuthHookG T.Text AuthHookType
+
+withAuthHook :: AuthHook -> (T.Text -> a) -> (T.Text -> a) -> a
+withAuthHook (AuthHookG u t) ifGet ifPost =
+  bool (ifGet u) (ifPost u) $ isPost t
+
+hookMethod :: AuthHook -> N.StdMethod
+hookMethod = bool N.GET N.POST . isPost . ahType
+
+
+-- | Extracts 'UserInfo' from the webhook answer.
+--   This uses the provided manager to make the call, extracts
+--   information such as response body and status, and delegates to a
+--   local 'mkUserInfoFromResp'.
+userInfoFromAuthHook
+  :: (HasVersion, MonadIO m, MonadError QErr m)
+  => Logger Hasura
+  -> H.Manager
+  -> AuthHook
+  -> [N.Header]
+  -> m (UserInfo, Maybe UTCTime)
+userInfoFromAuthHook logger manager hook reqHeaders = do
+  res  <- liftIO $ try $ withAuthHook hook withGET withPOST
+  resp <- either logAndThrow return res
+  let status   = resp ^. Wreq.responseStatus
+      respBody = resp ^. Wreq.responseBody
+  mkUserInfoFromResp logger (ahUrl hook) (hookMethod hook) status respBody
+
+  where
+    withGET  url = Wreq.getWith  (mkOptions filteredHeaders) $ T.unpack url
+    withPOST url = Wreq.postWith (mkOptions [contentType]) (T.unpack url) $
+      object ["headers" J..= postHdrsPayload]
+    mkOptions = wreqOptions manager
+    contentType = ("Content-Type", "application/json")
+    postHdrsPayload = J.toJSON $ Map.fromList $ hdrsToText reqHeaders
+
+    logAndThrow err = do
+      unLogger logger $
+        WebHookLog LevelError Nothing (ahUrl hook) (hookMethod hook)
+        (Just $ HttpException err) Nothing Nothing
+      throw500 "Internal Server Error"
+
+    filteredHeaders = flip filter reqHeaders $ \(n, _) ->
+      n `notElem` commonClientHeadersIgnored
```

Comment:
This is a style comment, so I will preface it with the disclaimer that this is subjective.

There is a Haskell style that seems to enjoy aggressively pulling out lots of expressions into `where` blocks (sometimes seemingly at random). Perhaps this is a little more aesthetically pleasing, but it harms readability: reading the function requires constantly bouncing around between the body and the `where` clause.

Moving nontrivial helper functions that represent meaningfully-distinct chunks of functionality into `where` clauses is great, and I wholeheartedly support doing that! But otherwise, I much prefer keeping related bits of logic close together in the code. Consider this rearranging of the logic in this function:

```haskell
userInfoFromAuthHook logger manager hook reqHeaders = do
  response <- (`onLeft` logAndThrow) =<< liftIO (try performHttpRequest)
  let status   = response ^. Wreq.responseStatus
      respBody = response ^. Wreq.responseBody
  mkUserInfoFromResponse logger (ahUrl hook) (hookMethod hook) status respBody
  where
    performHttpRequest = do
      let url = T.unpack $ ahUrl hook
      case ahType hook of
        AHTGet -> do
          let isCommonHeader header = (`elem` commonClientHeadersIgnored)
              filteredHeaders = filter (not . isCommonHeader . fst) reqHeaders
          Wreq.getWith (wreqOptions manager filteredHeaders) url
        AHTPost -> do
          let contentType = ("Content-Type", "application/json")
              headersPayload = Map.fromList $ hdrsToText reqHeaders
          Wreq.postWith (wreqOptions manager [contentType]) url $
            object ["headers" J..= headersPayload]
```

I’ve made the following changes:

  1. I moved most of the functionality into `performHttpRequest`, which means the function body is basically just wiring code that glues various pieces together.

  2. Since `performHttpRequest` is now a meaningful chunk of functionality, it can be given an equally-meaningful name.

  3. I’ve moved the auxiliary bindings into `let` blocks, defined as close to where they’re needed as possible. This makes it possible to read the code by just following the control flow.

Reply:
SGTM! I had kept the code pretty close to its original style, partly because, well, don't fix what ain't broke: code often is as it is for a good reason, and I didn't want to change too much of it just for style reasons. ^^

Reply:
Ah, you are right—that bit was part of the code that was moved, not something new! I was confused by the diff. It sounds like my argument was misplaced, so mea culpa!

Reply:
No, no, that's valid criticism! I had already made some small changes, I'll apply your suggestions too. :)

Reply:
Done!